### PR TITLE
BUG: disallow resample with non-Tick on TimedeltaIndex

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v2.1.0.rst [hunk 1]
```diff
@@ -225,6 +225,8 @@ Styler
 
 Other
 ^^^^^
+- Bug in :meth:`DataFrame.resample` and :meth:`Series.resample` in incorrectly allowing non-fixed ``freq`` when resampling on a :class:`TimedeltaIndex` (:issue:`51896`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:228]: Should probably go in `Groupby/resample/rolling`

File: pandas/core/resample.py [hunk 2]
```diff
@@ -1826,6 +1826,13 @@ def _get_time_delta_bins(self, ax: TimedeltaIndex):
                 f"an instance of {type(ax).__name__}"
             )
 
+        if not isinstance(self.freq, Tick):
```
Comment (mroeschke) [inline pandas/core/resample.py:1829]: What would resampling a timedeltaindex with a non-fixed frequency return before?

File: doc/source/whatsnew/v2.1.0.rst (hunk 1 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v2.1.0.rst]: moved+green

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**